xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java

```diff
@@ -0,0 +1,130 @@
+
+      // set attributes on the current object
+      for (final Field field : annotatedFields.getAttributeFields()) {
+        final String xmlAttributeValue = xmlStreamReader.getAttributeValue(null, field.getName());
+        final Object value = new AttributeValueCasting(field).castAttributeValue(xmlAttributeValue);
+        field.set(instance, value);
+      }
+
+      // Check if we have any more work to do, if only attributes can go ahead and return now.
+      if (annotatedFields.getTagFields().isEmpty()
+          && annotatedFields.getTagListFields().isEmpty()
+          && annotatedFields.getBodyTextFields().isEmpty()) {
+        return instance;
+      }
+
+      // This parser will do the work of parsing the current tag, it'll look at all
+      // child tags and the body text and invoke the right callback that we will define below.
+      final XmlParser tagParser = new XmlParser(pojo.getSimpleName());
+
+      // Set up tag parsing, as we scan through more elements when we see a matching
+      // tag name we'll call the child tag handler. The child tag handler will
+      // create a java model representing the child tag and set the field instance
+      // on our current running instance object.
+      for (final Field field : annotatedFields.getTagFields()) {
+        final String expectedTagName = field.getType().getSimpleName();
+        tagParser.childTagHandler(
+            expectedTagName, () -> field.set(instance, mapXmlToObject(field.getType())));
+      }
+
+      // Set up tag list parsing, similar to tag parsing except we set the field
+      // value to a list and each time we see a new child tag we'll add it back to that list.
+      for (final Field field : annotatedFields.getTagListFields()) {
+        final List<Object> tagList = new ArrayList<>();
+        field.set(instance, tagList);
+        final Class<?> listType = ReflectionUtils.getGenericType(field);
+        tagParser.childTagHandler(
+            listType.getSimpleName(), () -> tagList.add(mapXmlToObject(listType)));
+      }
+
+      // Set up body text handler. The XML cursor will iterate over each line of body
+      // content and we will buffer that content, when the full content is read it is trimmed
+      // and the setter callback below is executed.
+      for (final Field field : annotatedFields.getBodyTextFields()) {
+        Preconditions.checkState(annotatedFields.getBodyTextFields().size() == 1);
+        tagParser.bodyHandler(
+            textContent -> {
+              try {
+                field.set(instance, textContent);
+              } catch (final IllegalAccessException e) {
+                throw new JavaDataModelException(field, "Unexpected illegal access", e);
+              }
+            });
+      }
+
+      tagParser.parse(xmlStreamReader);
+
+      return instance;
+    } catch (final Throwable e) {
+      if (e instanceof XmlParsingException) {
```

Comment:
Codacy found an issue: [An instanceof check is being performed on the caught exception.  Create a separate catch clause for this exception type.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)